impl/src/main/java/org/glassfish/soteria/authorization/spi/impl/SubjectParser.java

```diff
@@ -490,11 +495,26 @@ private Principal doGetCallerPrincipalFromPrincipals(Iterable<Principal> princip
                         } catch (Exception e) {
                             
                         }
-                        break;
-                // TODO: depend on this directly later 
-                case "javax.security.enterprise.CallerPrincipal":
-                    return principal;
+                    break;
+            }
+
+            if (CallerPrincipal.class.isAssignableFrom(principal.getClass())) {
+                return principal;
+            }
+
+            // Check first if we are inside an EJB as it is more specific than a Servlet 
+            EJBContext ejbContext = EJB.getEJBContext();
+            if (ejbContext != null) {
+                return ejbContext.getCallerPrincipal();
```

Comment:
This code needs to account for the case that EJB.getEJBContext().getCallerPrincipal() returns a principal with an empty/null name (i.e., EJB's representation of an anonymous/unauthenticated users), and return null instead of a principal in that case.